src/Shared/ChangeWaves.cs

```diff
@@ -0,0 +1,204 @@
+{
+    internal enum ChangeWaveConversionState
+    {
+        NotConvertedYet,
+        Valid,
+        InvalidFormat,
+        OutOfRotation,
+        InvalidVersion
+    }
+
+    /// <summary>
+    /// All waves are enabled by default, meaning all features behind change wave versions are enabled.
+    /// </summary>
+    public class ChangeWaves
+    {
+        public static readonly string[] AllWaves = { Wave16_8, Wave16_10, Wave17_0 };
+        public static readonly Version[] AllWavesAsVersion = Array.ConvertAll<string, Version>(AllWaves, Version.Parse);
+        public const string Wave16_8 = "16.8";
+        public const string Wave16_10 = "16.10";
+        public const string Wave17_0 = "17.0";
+
+        /// <summary>
+        /// Special value indicating that all features behind change-waves should be enabled.
+        /// </summary>
+        public const string EnableAllFeatures = "999.999";
+
+        internal static readonly Version LowestWaveAsVersion = new Version(AllWaves[0]);
+        internal static readonly Version HighestWaveAsVersion = new Version(AllWaves[AllWaves.Length - 1]);
+        internal static readonly Version EnableAllFeaturesAsVersion = new Version(EnableAllFeatures);
+
+        internal static string LowestWave
+        {
+            get
+            {
+                return AllWaves[0];
+            }
+        }
+
+        internal static string HighestWave
+        {
+            get
+            {
+                return AllWaves[AllWaves.Length - 1];
+            }
+        }
+
+        private static string cachedWave = null;
+
+        public static string DisabledWave
+        {
+            get
+            {
+                if (cachedWave == null)
+                {
+                    cachedWave = Traits.Instance.MSBuildDisableFeaturesFromVersion ?? ChangeWaves.EnableAllFeatures;
+                }
+
+                return cachedWave;
+            }
+            set
+            {
+                cachedWave = value;
+            }
+        }
+
+        private static ChangeWaveConversionState _state;
+        internal static ChangeWaveConversionState ConversionState
+        {
+            get
+            {
+                return _state;
+            }
+            set
+            {
+                // Keep state persistent.
+                if (_state == ChangeWaveConversionState.NotConvertedYet)
+                {
+                    _state = value;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Ensure the the environment variable MSBuildDisableFeaturesFromWave is set to a proper value.
+        /// </summary>
+        /// <returns> String representation of the set change wave. "999.999" if unset or invalid, and clamped if out of bounds. </returns>
+        internal static void ApplyChangeWave()
+        {
+            Version changeWave;
+
+            // If unset, enable all features.
+            if (DisabledWave.Length == 0)
```

Comment:
nit:
convert to switch statement and reorder according to which you think will be used most often. (So EnableAllFeaturesAsVersion at top, for instance.

Reply:
A switch statement based on what? I do agree this can be implemented better.